data/blog/2022-08-01---10-things.md

```diff
@@ -0,0 +1,144 @@
+
+I was looking in to [the implementation](https://github.com/python/cpython/blob/main/Lib/functools.py#L566) of lru_cache in the CPython repos and I figured I could beat it. Spoiler: I couldn't. It turns out there was a [C version](https://github.com/python/cpython/blob/main/Modules/_functoolsmodule.c#L992) which makes it very fast for both cache hits and cache misses.
+
+Knowing it is that fast convinced me to lower the barrier to use `@lru_cache`. There are a number of small functions in Textual, that are not exactly slow, but called a large number of times. Many of them were highly cacheable and judicious use of `@lru_cache` provided a significant win. Typically a `maxsize` of around 1000-4000 was enough to ensure that the majority calls were cache hits.
+
+Here's an example of the kind of function that benefited from caching. This method combines two rectangular regions in to a single region that fits both. You can see it doesn't do a great deal of work, but it was called 1000s of times.
+
+<script src="https://gist.github.com/willmcgugan/8d161a6b1104aa2b79b6656893d4055d.js"></script>
+
+A word of advice when using `lru_cache`: always check your assumptions by inspecting the `cache_info()`. For effective caching you should expect to see `hits` growing faster than `misses`.
+
+### Immutable is best
+
+Following on from the previous tip, I'd like to sing the praises of immutable objects. Python doesn't have true immutable objects, but you can get much of the benefits from tuples, NamedTuples, or frozen dataclasses.
+
+It _seems_ like an arbitrary limitation that you can't change an object, but it rarely is in practice. Computer scientists will point out that many languages only have immutable objects.
+
+In Textual, the code that uses immutable objects is the easiest to reason about, easiest to cache, and easiest to test. Mainly because you can write code that is free of side-effects. Something that is difficult to do when you pass class instances in to a function.
+
+### Unicode art is good
+
+Some technical things are hard to explain in words, and a diagram created from unicode box characters can be massively beneficial in documentation. This diagram is taken from a docstring in Textual for a method that splits a region in to four sub-regions:
+
+```
+           cut_x ↓
+        ┌────────┐ ┌───┐
+        │        │ │   │
+        │    0   │ │ 1 │
+        │        │ │   │
+cut_y → └────────┘ └───┘
+        ┌────────┐ ┌───┐
+        │    2   │ │ 3 │
+        └────────┘ └───┘
+```
+
+It's no substitute for a well written docstring, but in combination it is super-helpful. I'd encourage you to add diagrams to docstrings wherever it makes sense.
+
+I use [monodraw](https://monodraw.helftone.com/) for these diagrams. Monodraw is MacOS only unfortunately, but there are no doubt good alternatives for other platforms.
```

Comment:
Perhaps someone will build a cross-platform Monodraw competitor in Textual?

Reply:
That would be awesome.

Reply:
Yeah, would be a great use case for it!